docs/fields.rst

```diff
@@ -446,54 +446,146 @@ Simple choices do not have to explicitly use vocabularies:
+   ...
+   InvalidVocabularyError: Invalid vocabulary <function myDynamicVocabulary at 0x1101f7730>
+   >>> f = f.bind(3)
+   >>> f.validate(1)
+   >>> f.validate(2)
+   >>> f.validate(3)
+   Traceback (most recent call last):
+   ...
+   ConstraintNotSatisfied: 3
+
+Named (Registered) Vocabularies
+-------------------------------
+
+We can also provide a vocabulary name that will be resolved later
+against a registry of vocabulary factories (objects that implement
+:class:`zope.schema.interfaces.IVocabularyFactory`). On the surface,
+this looks very similar to providing a ``source`` argument: they are
+both callable objects that take a context and return a vocabulary. The
+advantage of a named factory is a level of indirection, allowing the
+same name to be easily used in many different fields, even from
+packages that aren't aware of each other. For example, an application
+framework may define choices that use a 'permissions' vocabulary, and
+individual applications may define their own meaning for that name.
+
+A simple version of this is provided in this package using a global
+vocabulary registry:
 
 .. doctest::
 
    >>> from zope.schema.vocabulary import SimpleVocabulary
-   >>> def myDynamicVocabulary(context):
-   ...     v = dynamic_context_calculation_that_returns_an_iterable(context)
-   ...     return SimpleVocabulary.fromValues(v)
+   >>> from zope.schema.vocabulary import getVocabularyRegistry
+   >>> from zope.schema.interfaces import IVocabularyFactory
+   >>> from zope.interface import implementer
+   >>> @implementer(IVocabularyFactory)
+   ... class PermissionsVocabulary(object):
    ...
+   ...     def __call__(self, context):
+   ...         if context is None: raise AttributeError
+   ...         return SimpleVocabulary.fromValues(context.possible_permissions)
+   >>> getVocabularyRegistry().register('permissions', PermissionsVocabulary())
+
+   >>> class Context(object):
+   ...     possible_permissions = ('read', 'write')
+
+Unlike ``IContextSourceBinder``, the factory is invoked even for
+unbound fields; depending on the factory, this may or may not do
+anything useful (our factory produces errors):
 
-The vocabulary interface is simple enough that writing a custom vocabulary is
-not too difficult itself.
+.. doctest::
+
+   >>> f = Choice(vocabulary='permissions')
+   >>> f.validate('read')
+   Traceback (most recent call last):
+   ...
+   AttributeError
+   >>> context = Context()
+   >>> f = f.bind(context)
+   >>> f.validate("read")
+   >>> f.validate("write")
+   >>> f.validate("delete")
+   Traceback (most recent call last):
+   ...
+   ConstraintNotSatisfied: ('delete', '')
 
-See zope.schema.vocabulary.TreeVocabulary for another
-IBaseVocabulary supporting vocabulary that provides a nested, tree-like
-structure.
+A registry that keeps factories as named utilities in the `Zope
+component architecture <https://zopecomponent.readthedocs.io>`_ is
+provided by the `zope.vocabularyregistry
```

Comment:
Active voice here, please.

Reply:
I think I was trying to keep a similar style to the sentence you objected to above 😄